### ISSUE TITLE
Doesn't rerender when props change

### ISSUE THREAD

**Issue by matthewp** [skatejs/skatejs#1134]
I think [this line](https://github.com/skatejs/skatejs/blob/dac2c4dff7824221aebf3bf55c402c04b0ad8674/src/with-props.js#L92) is wrong.  It should be `!keys(prev).every(k => prev[k] === next[k])`.

Going to try my hand at writing a test for this :D

**Comment by matthewp** [comment]
Even though (I think), that's wrong, regular components are fine because with-render.js ignores that: https://github.com/skatejs/skatejs/blob/dac2c4dff7824221aebf3bf55c402c04b0ad8674/src/with-render.js#L13

I don't quite understand, is the point of propsUpdatedCallback to determine if the props have changed? If so, shouldn't with-render be using propsChangedCallback instead?  @treshugart

**Comment by Hotell** [comment]
> I think this line is wrong. It should be !keys(prev).every(k => prev[k] === next[k]).

I think you are right @matthewp , but it makes it very non human readable and also iterating through all props may be slow.

This should do the trick:

```
keys(prev).some(k => prev[k] !== next[k])
```

**Comment by Hotell** [comment]
> Even though (I think), that's wrong, regular components are fine because with-render.js ignores that: https://github.com/skatejs/skatejs/blob/dac2c4dff7824221aebf3bf55c402c04b0ad8674/src/with-render.js#L13

Hmm good catch! I think this is wrong, because if props have been not changed we don't wanna trigger render, otherwise it don't make sense to have it there at all

**Comment by treshugart** [comment]
@Hotell I would slightly amend that to:

```js
!prev || keys(prev).some(k => prev[k] !== next[k]);
```

We know it will be `undefined` or an object, so shortening it is fine.

**Comment by treshugart** [comment]
@matthewp it looks like we should be using `propsChangedCallback` instead as that will only be called when `propsUpdatedCallback` returns true, and we're not overriding it.

**Comment by treshugart** [comment]
I'm also happy to change names to reflect the intent better, if there are any clearly better suggestions. I don't want to spend too much time on that bike shed, though.

**Comment by matthewp** [comment]
I don't want to bikeshed either, was just trying to understand the code.  I was mostly wondering if with-render used propsUpdatedCallback on purpose for a reason or now.

Ok, so then I think the conclusion here is that:

1. with-render should use `propsChangedCallback ` and not override `propsUpdatedCallback`.
2. Fix the bug in with-props.

I think I can handle this!

**Comment by treshugart** [comment]
Yeah, that sounds great! Thanks, @matthewp.

**Comment by matthewp** [comment]
There's a problem with the way we render. This doesn't do what we want in Preact:

```js
var div = document.createElement('div');
div.attachShadow({mode:'open'});
document.body.appendChild(div);

render(h('div', null, 'foo'), div.shadowRoot);
render(h('div', null, 'bar'), div.shadowRoot);
```

It will create 2 divs, not overwrite the first one. What is the correct way to re-render an element with vdom, without creating a preact component. Or should we create a preact component?

**Comment by matthewp** [comment]
Answered on twitter :D

**Comment by matthewp** [comment]
https://github.com/skatejs/skatejs/pull/1135

**Comment by matthewp** [comment]
This was fixed by #1135
